moshi/src/main/java/com/squareup/moshi/internal/Util.kt

```diff
@@ -0,0 +1,643 @@
+    is Class<*> -> {
+      if (isArray) GenericArrayTypeImpl(this@canonicalize.componentType.canonicalize()) else this
+    }
+    is ParameterizedType -> {
+      if (this is ParameterizedTypeImpl) return this
+      ParameterizedTypeImpl(ownerType, rawType, *actualTypeArguments)
+    }
+    is GenericArrayType -> {
+      if (this is GenericArrayTypeImpl) return this
+      GenericArrayTypeImpl(genericComponentType)
+    }
+    is WildcardType -> {
+      if (this is WildcardTypeImpl) return this
+      WildcardTypeImpl(upperBounds, lowerBounds)
+    }
+    else -> this // This type is unsupported!
+  }
+}
+
+/** If type is a "? extends X" wildcard, returns X; otherwise returns type unchanged. */
+public fun Type.removeSubtypeWildcard(): Type {
+  if (this !is WildcardType) return this
+  val lowerBounds = lowerBounds
+  if (lowerBounds.isNotEmpty()) return this
+  val upperBounds = upperBounds
+  require(upperBounds.size == 1)
+  return upperBounds[0]
+}
+
+public fun Type.resolve(context: Type, contextRawType: Class<*>): Type {
+  return this.resolve(context, contextRawType, LinkedHashSet())
+}
+
+private fun Type.resolve(
+  context: Type,
+  contextRawType: Class<*>,
+  visitedTypeVariables: MutableCollection<TypeVariable<*>>
+): Type {
+  // This implementation is made a little more complicated in an attempt to avoid object-creation.
+  var toResolve = this
+  while (true) {
+    when {
+      toResolve is TypeVariable<*> -> {
+        val typeVariable = toResolve
+        if (typeVariable in visitedTypeVariables) {
+          // cannot reduce due to infinite recursion
+          return toResolve
+        } else {
+          visitedTypeVariables += typeVariable
+        }
+        toResolve = resolveTypeVariable(context, contextRawType, typeVariable)
+        if (toResolve === typeVariable) return toResolve
+      }
+      toResolve is Class<*> && toResolve.isArray -> {
+        val original = toResolve
+        val componentType: Type = original.componentType
+        val newComponentType = componentType.resolve(context, contextRawType, visitedTypeVariables)
+        return if (componentType === newComponentType) original else newComponentType.asArrayType()
+      }
+      toResolve is GenericArrayType -> {
+        val original = toResolve
+        val componentType = original.genericComponentType
+        val newComponentType = componentType.resolve(context, contextRawType, visitedTypeVariables)
+        return if (componentType === newComponentType) original else newComponentType.asArrayType()
+      }
+      toResolve is ParameterizedType -> {
+        val original = toResolve
+        val ownerType: Type? = original.ownerType
+        val newOwnerType = ownerType?.let {
+          ownerType.resolve(context, contextRawType, visitedTypeVariables)
+        }
+        var changed = newOwnerType !== ownerType
+        var args = original.actualTypeArguments
+        var t = 0
+        val length = args.size
+        while (t < length) {
```

Comment:
isn't this equivalent, or am I missing something
```kotlin
for(t in 0..args.size)
```

Reply:
Close! `for(t in args.indices) {`